ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/DbChangelog.hs

```diff
@@ -0,0 +1,428 @@
+prop_flushDbChangelogKeepsInvariants setup =
+  property $ checkInvariants toFlush && checkInvariants toKeep
+  where
+    (toFlush, toKeep) = flushDbChangelog DbChangelogFlushAllImmutable $
+      resultingDbChangelog setup
+
+prop_rollbackDbChangelogKeepsInvariants ::
+  DbChangelogTestSetupWithRollbacks -> Property
+prop_rollbackDbChangelogKeepsInvariants setup = property $ checkInvariants dblog
+  where
+    n = rollbacks setup
+    dblog = rollbackDbChangelog n (resultingDbChangelog $ testSetup setup)
+
+prop_prefixBackToAnchorIsRollingBackVolatileStates :: DbChangelogTestSetup -> Property
+prop_prefixBackToAnchorIsRollingBackVolatileStates setup =
+  property $ rolledBack == toAnchor
+  where
+    dblog = resultingDbChangelog setup
+    n = AS.length (changelogVolatileStates dblog)
+    rolledBack = rollbackDbChangelog n dblog
+    toAnchor = prefixBackToAnchorDbChangelog dblog
+
+prop_rollBackToVolatileTipIsNoop ::
+  Positive Int -> DbChangelogTestSetup -> Property
+prop_rollBackToVolatileTipIsNoop (Positive n) setup = property $ Just dblog == dblog'
+  where
+    dblog = resultingDbChangelog setup
+    pt = getTip $ unDbChangelogState $ AS.headAnchor $ changelogVolatileStates dblog
+    dblog' = prefixDbChangelog pt $ nExtensions n dblog
+
+unsafeJoinSeqUtxoDiffs :: Ord k => SeqUtxoDiff k v -> SeqUtxoDiff k v -> SeqUtxoDiff k v
+unsafeJoinSeqUtxoDiffs (SeqUtxoDiff ft1) (SeqUtxoDiff ft2) =
+  SeqUtxoDiff (ft1 FT.>< ft2)
+
+
+{-------------------------------------------------------------------------------
+  Generators
+-------------------------------------------------------------------------------}
+
+pointAtSlot :: WithOrigin SlotNo -> Point (TestLedger EmptyMK)
+pointAtSlot = Point.withOrigin GenesisPoint (\slotNo -> Point $ At $ Point.Block slotNo H)
+
+type Key = String
+
+data GenOperationsState = GenOperationsState {
+    gosSlotNo            :: !(WithOrigin SlotNo)
+  , gosOps               :: ![Operation TestLedger]
+  , gosActiveUtxos       :: !(Map Key Int)
+  , gosPendingInsertions :: !(Map Key Int)
+  , gosConsumedUtxos     :: !(Set Key)
+  } deriving (Show)
+
+applyPending :: GenOperationsState -> GenOperationsState
+applyPending gosState = gosState
+  { gosActiveUtxos = Map.union (gosActiveUtxos gosState) (gosPendingInsertions gosState)
+  , gosPendingInsertions = Map.empty
+  }
+
+genOperations :: WithOrigin SlotNo -> Int -> Gen [Operation TestLedger]
+genOperations slotNo nOps = gosOps <$> execStateT (replicateM_ nOps genOperation) initState
+  where
+    initState = GenOperationsState {
+        gosSlotNo = slotNo
+      , gosActiveUtxos = Map.empty
+      , gosPendingInsertions = Map.empty
+      , gosConsumedUtxos = Set.empty
+      , gosOps = []
+      }
+
+    genOperation = do
+      op <- frequency' [ (1, genPrune), (10, genExtend) ]
+      modify' $ \st -> st { gosOps = op:gosOps st }
+
+    genPrune = Prune <$> SecurityParam <$> lift (chooseEnum (1, 10))
```

Comment:
We use various ranges for generating arbitrary instances, like 1 - 10 for security param above, or choosing next slot incrementally within 1-5 range below and few more. Would it make sense to organise them together for convenience, understanding and later might be even used for running tests with different parameters. Just an observation! @bartfrenk  and @dnadales